### PR TITLE
Compile snappy, lz4, exml using port_compiler rebar plugin

### PR DESCRIPTION
**Pull request by arcusfelis**
Decided to make PRs to external libraries instead:
- [x] Open PR for fast_tls with the changes: https://github.com/processone/fast_tls/pull/38
- [x] Open PR for exml https://github.com/esl/exml/pull/46
- [x] Open PR for snappy (dependency of cqerl) https://github.com/skunkwerks/snappy-erlang-nif/pull/26 
- [x] Open PR for lz4 (dependency of cqerl) - https://github.com/szktty/erlang-lz4/pull/10
- [x] Wait for exml PR to be merged
- [x] Wait for snappy PR to be merged
- [x] Wait for lz4 PR to be merged
- [x] Update cqerl with new versions of deps lz4 and snappy https://github.com/esl/cqerl/pull/6
- [x] Update cqerl for MongooseIM
- [x] Update fast_tls for MongooseIM
- [x] Tag exml 3.0.3 https://github.com/esl/exml/releases/tag/3.0.3
- [x] Update exml to 3.0.3 for MongooseIM
- [ ] Use Rebar 3 to compile escalus (or rebar 2.6.4). We've discussed it in https://github.com/esl/exml/pull/47 issue
- [ ] Update exml to 3.0.3 for escalus https://github.com/esl/escalus/pull/192
- [ ] Update escalus in MongooseIM's big tests
- [ ] Upgrade fast_tls for big tests (used by stun, overridden)

Old description:

Instead of Makefiles, that have issues with escaping and the way they get paths from erlang
(they require to disable progress report from sasl at least)

Fix fast_tls compilation with ERTS installed in a directory with a whitespace

This PR addresses:
- can't compile MongooseIM, if ERTS is installed into a directory with whitespaces
- makefile does not filter out progress reports:

```erlang
erl -noshell -noinput -s init stop -eval "io:format(\"~s/erts-~s/include/\", [code:root_dir(), erlang:system_info(version)])."

    =PROGRESS REPORT==== 5-Nov-2018::14:39:59 ===
              supervisor: {local,sasl_safe_sup}
                 started: [{pid,<0.56.0>},
                           {id,alarm_handler},
                           {mfargs,{alarm_handler,start_link,[]}},
                           {restart_type,permanent},
                           {shutdown,2000},
                           {child_type,worker}]
    
    =PROGRESS REPORT==== 5-Nov-2018::14:39:59 ===
              supervisor: {local,sasl_sup}
                 started: [{pid,<0.55.0>},
                           {id,sasl_safe_sup},
                           {mfargs,
                               {supervisor,start_link,
                                   [{local,sasl_safe_sup},sasl,safe]}},
                           {restart_type,permanent},
                           {shutdown,infinity},
                           {child_type,supervisor}]
    
    =PROGRESS REPORT==== 5-Nov-2018::14:39:59 ===
              supervisor: {local,sasl_sup}
                 started: [{pid,<0.57.0>},
                           {id,release_handler},
                           {mfargs,{release_handler,start_link,[]}},
                           {restart_type,permanent},
                           {shutdown,2000},
                           {child_type,worker}]
    
    =PROGRESS REPORT==== 5-Nov-2018::14:39:59 ===
             application: sasl
              started_at: nonode@nohost
    /Users/mikhailuvarov/Library/Application Support/ErlangInstaller/19.3/erts-8.3/include/
```

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -157,6 +157,62 @@
...
+                     "priv/lz4_nif.so",
+                     ["c_src/*.c"],
+                     []
+                    }]},
+      {provider_hooks,
+       [ {post, [ {compile, {pc, compile}},
+                  {clean, {pc, clean}} ] }] } ]},
+
+   %% Compile with port_compiler instead of Makefile
+   {override, exml,
```
Comment (michalwski) [inline rebar.config]: Could it be done in exml directly instead of the override here?

Comment (arcusfelis) [reply rebar.config]: yep, this PR is more like a prototype.
same with fast_tls part, we should just make PR.

But I am not sure if lz4 and snappy maintainers are happy with pc...

Comment (michalwski) [reply rebar.config]: > But I am not sure if lz4 and snappy maintainers are happy with pc...

Did you ask them?

Comment (arcusfelis) [reply rebar.config]: nope. I can make PRs to all the libs and check...

Comment (michalwski) [reply rebar.config]: Please do :)

**Comment by arcusfelis** [comment]
PR for fast_tls with the changes: https://github.com/processone/fast_tls/pull/38

**Comment by mongoose-im** [comment]
[5826.1](https://travis-ci.org/esl/MongooseIM/jobs/451400309) / Erlang 19.3 / small_tests / cb289e8af482aea79d83fb123152a902cbb4a351
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2124/5826/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2124/5826/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-06_16.59.51/lib.mongooseim.logs/run.2018-11-06_16.59.51/suite.log.html)

**Comment by arcusfelis** [comment]
<details><summary>Original diff inside...</summary>
<p>

```diff
diff --git a/rebar.config b/rebar.config
index c680c99a0..87c9a575a 100644
--- a/rebar.config
+++ b/rebar.config
@@ -157,6 +157,62 @@
     [ {provider_hooks,
        [ {post, [ {compile, {pc, compile}},
                   {clean, {pc, clean}} ] }] } ]},
+
+   %% Compile with port_compiler instead of Makefile
+   {override, snappy,
+    [ {pre_hooks, []},
+      {post_hooks, []},
+      {plugins, [pc]},
+      {artifacts, ["priv/snappy_nif.so"]},
+      {port_specs, [{".*",
+                     "priv/snappy_nif.so",
+                     ["c_src/*.cc"],
+                     []
+                    }]},
+      {provider_hooks,
+       [ {post, [ {compile, {pc, compile}},
+                  {clean, {pc, clean}} ] }] } ]},
+
+   %% Compile with port_compiler instead of Makefile
+   {override, lz4,
+    [ {pre_hooks, []},
+      {post_hooks, []},
+      {plugins, [pc]},
+      {artifacts, ["priv/lz4_nif.so"]},
+      {port_specs, [{".*",
+                     "priv/lz4_nif.so",
+                     ["c_src/*.c"],
+                     []
+                    }]},
+      {provider_hooks,
+       [ {post, [ {compile, {pc, compile}},
+                  {clean, {pc, clean}} ] }] } ]},
+

+   %% Compile with port_compiler instead of Makefile
+   {override, exml,
+    [ {pre_hooks, []},
+      {post_hooks, []},
+      {plugins, [pc]},
+      {artifacts, ["priv/exml_nif.so"]},
+      {port_specs, [{".*",
+                     "priv/exml_nif.so",
+                     ["c_src/*.cpp"],
+                     [ {env, [{"CXXFLAGS", "$CXXFLAGS -std=c++11"}]} ]
+                    }]},
+      {provider_hooks,
+       [ {post, [ {compile, {pc, compile}} ]},
+         {post, [ {clean, {pc, clean}} ]}
+       ]}
+    ]},
+
+   %% BINDIR is not properly escaped, which breaks compilation in case
+   %% Erlang is installed into a directory with whitespaces
+   %% (i.e. with ErlangInstaller, which puts erlang into
+   %% /Users/mikhailuvarov/Library/Application Support/ErlangInstaller/21.1/)
+   {override, fast_tls,
+    [ {port_env, [{"CFLAGS", "$CFLAGS"}, {"LDFLAGS", "$LDFLAGS -lssl -lcrypto"},
+            {"darwin", "DRV_LDFLAGS", "-bundle -bundle_loader \"${BINDIR}/beam.smp\" $ERL_LDFLAGS"}]} ]},
+
    %% Disable unsupported plugins for hamcrest-erlang
    {override, riakc, [{erl_opts, [debug_info,
                                   {d, namespaced_types},
```

</p>
</details>

File: rebar.config (hunk 1 above)
Comment (dch) [reply rebar.config]: BTW if there's interest from ESL to be added as maintainers for snappy just drop me an email.

**Comment by mongoose-im** [comment]
[5842.1](https://travis-ci.org/esl/MongooseIM/jobs/453651161) / Erlang 19.3 / small_tests / 7d813980c810179d1bd4df274421ad6884620d8d
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2124/5842/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2124/5842/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-11_18.58.39/lib.mongooseim.logs/run.2018-11-11_18.58.40/suite.log.html)

**Comment by mongoose-im** [comment]
[5843.1](https://travis-ci.org/esl/MongooseIM/jobs/453651736) / Erlang 19.3 / small_tests / 4816300883e802680769381d5d1587e9dd3d9746
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2124/5843/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2124/5843/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-11_19.32.52/lib.mongooseim.logs/run.2018-11-11_19.32.53/suite.log.html)

**Pull request closed without merging**